### PR TITLE
VAULT-19681 allow users to specify files for agent child process stdout/stderr

### PR DESCRIPTION
**Pull request by dhuckins**
Adds ability for the user to send stderr and stdout of the child process to files

### CODE REVIEW

File: command/agent.go [hunk 1]
```diff
@@ -729,13 +729,17 @@ func (c *AgentCommand) Run(args []string) int {
...
-		es := exec.NewServer(&exec.ServerConfig{
+		es, err := exec.NewServer(&exec.ServerConfig{
 			AgentConfig: c.config,
 			Namespace:   templateNamespace,
 			Logger:      c.logger.Named("exec.server"),
 			LogLevel:    c.logger.GetLevel(),
 			LogWriter:   c.logWriter,
 		})
+		if err != nil {
+			c.logger.Error("could not create exec server", "error", err)
```
Comment (VioletHynes) [inline command/agent.go:740]: good catch!

File: command/agent.go [hunk 2]
```diff
@@ -800,6 +804,7 @@ func (c *AgentCommand) Run(args []string) int {
 				leaseCache.SetShuttingDown(true)
 			}
 			cancelFunc()
+			es.Close()
```
Comment (VioletHynes) [inline command/agent.go:807]: Does this need an `if es != nil` or would that be overly defensive?

File: command/agent/exec/exec.go [hunk 3]
```diff
@@ -333,3 +367,19 @@ func (s *Server) restartChildProcess(newEnvVars []string) error {
...
 }
+
+func (s *Server) Close() {
+	s.childProcessLock.Lock()
+	defer s.childProcessLock.Unlock()
+	s.close()
+}
+
+func (s *Server) close() {
+	if s.childProcessCloseStdOut {
```
Comment (VioletHynes) [inline command/agent/exec/exec.go:378]: Out of curiosity -- why do we need this? I checked the documentation and it doesn't look like there'd be any issue calling `os.Stdout.Close()`, but I'd love to hear more!

File: command/agent/exec/exec_test.go [hunk 4]
```diff
@@ -380,3 +385,164 @@ func TestExecServer_Run(t *testing.T) {
 		})
 	}
 }
+
+func TestExecServer_LogFiles(t *testing.T) {
```
Comment (VioletHynes) [inline command/agent/exec/exec_test.go:389]: From what I can tell, this only tests the `stderr` case -- we should add a test for `stdout` too.

File: changelog/22812.txt
Comment (VioletHynes) [inline changelog/22812.txt:1]: We'll need docs for this feature too -- will they be coming in a different PR, or should I expect them to be added to this one?

File: command/agent/exec/exec_test.go [hunk 5]
```diff
@@ -380,3 +385,164 @@ func TestExecServer_Run(t *testing.T) {
...
+		expectedError error
+	}{
+		"can_log_to_file": {
+			testAppPort: 34001,
+			stderrFile:  tempStderr,
+		},
+		"cant_open_file": {
+			testAppPort:   34002,
+			stderrFile:    "/file/does/not/exist",
+			expectedError: os.ErrNotExist,
```
Comment (VioletHynes) [inline command/agent/exec/exec_test.go:423]: Why do we error if the file doesn't exist? We open it with `os.O_APPEND|os.O_CREATE|os.O_WRONLY`, so shouldn't be be able to handle writing to a new file?

File: command/agent/config/config.go [hunk 6]
```diff
@@ -171,6 +171,8 @@ type ExecConfig struct {
 	Command                []string  `hcl:"command,attr" mapstructure:"command"`
 	RestartOnSecretChanges string    `hcl:"restart_on_secret_changes,optional" mapstructure:"restart_on_secret_changes"`
 	RestartStopSignal      os.Signal `hcl:"-" mapstructure:"restart_stop_signal"`
+	ChildProcessStdout     string    `mapstructure:"child_process_stdout"`
+	ChildProcessStderr     string    `mapstructure:"child_process_stderr"`
```
Comment (averche) [inline command/agent/config/config.go:175]: Should we add `hcl` tags here as well?

File: command/agent/exec/exec.go [hunk 7]
```diff
@@ -65,9 +65,13 @@ type Server struct {
...
 
-	childProcess      *child.Child
-	childProcessState childProcessState
-	childProcessLock  sync.Mutex
+	childProcess            *child.Child
+	childProcessState       childProcessState
+	childProcessLock        sync.Mutex
+	childProcessStdout      io.WriteCloser
+	childProcessStderr      io.WriteCloser
+	childProcessCloseStdOut bool
```
Comment (averche) [inline command/agent/exec/exec.go:73]:
```suggestion
	childProcessCloseStdout bool
```

File: command/agent/exec/exec_test.go [hunk 8]
```diff
@@ -380,3 +385,164 @@ func TestExecServer_Run(t *testing.T) {
...
+				execServerTokenCh = make(chan string, 1)
+			)
+			go func() {
+				execServerErrCh <- execServer.Run(ctx, execServerTokenCh)
+			}()
+
+			// send a dummy token to kick off the server
+			execServerTokenCh <- "my-token"
+
+			// ensure the test app is running after 3 seconds
```
Comment (averche) [inline command/agent/exec/exec_test.go:497]:
```suggestion
			// ensure the test app is running after 500ms
```

**Review by averche — COMMENTED** [review]
Looks reasonable to me. Could you please add "agent" to the title/description.

File: command/agent/exec/exec_test.go (hunk 5 above)
Comment (dhuckins) [reply command/agent/exec/exec_test.go:423]: its if the parent folder doesn't exist

File: changelog/22812.txt
Comment (dhuckins) [reply changelog/22812.txt:1]: different PR

File: command/agent/exec/exec.go (hunk 3 above)
Comment (dhuckins) [reply command/agent/exec/exec.go:378]: I'm honestly not sure... I know not to close stdout with python, but not sure with go

would have to look up the order of things, wouldn't want to close stdout then if something (like consul-template for example) tries to write to it

Comment (VioletHynes) [reply command/agent/exec/exec.go:378]: It would be nice if we could avoid this additional complexity, so I think it would be nice if we could do it without the `childProcessCloseStdOut` yeah. From what I can tell, nothing seems to be wrong with closing stdout in go, and we'll be closing the process anyway, right? So there's no situation where we'd close it and another process will need it, right?

File: command/agent.go (hunk 2 above)
Comment (dhuckins) [reply command/agent.go:807]: think it would be overly defensive, `es` would only be `nil` if there was an error opening one of the log files and in that case would exit early

or should we check anyway b/c its in a closure?

File: command/agent/exec/exec.go (hunk 3 above)
Comment (dhuckins) [reply command/agent/exec/exec.go]: correct! will try it (looks a lot simpler this way) thanks!

File: command/agent.go (hunk 2 above)
Comment (VioletHynes) [reply command/agent.go:807]: I'll leave it up to you -- I think it wouldn't hurt, but like you said, I can't really see how we'd get to that point

**Review by VioletHynes — APPROVED** [review]
Looks good! Thanks for this! Feel free to tag me on the docs PR, too!

File: command/agent/exec/exec.go [hunk 9]
```diff
@@ -333,3 +359,14 @@ func (s *Server) restartChildProcess(newEnvVars []string) error {
...
+
+func (s *Server) Close() {
+	s.childProcessLock.Lock()
+	defer s.childProcessLock.Unlock()
+	s.close()
+}
+
+func (s *Server) close() {
+	_ = s.childProcessStdout.Close()
+	_ = s.childProcessStderr.Close()
```
Comment (averche) [inline command/agent/exec/exec.go:371]: Would this potentially close the global os.Stdout and os.Stderr? https://go.dev/play/p/ksHYajR4CVU

**Review by averche — APPROVED** [review]
LGTM 👍
